aten/src/ATen/native/vulkan/api/Resource.h

```diff
@@ -25,12 +25,25 @@ struct C10_EXPORT Resource final {
     template<
         typename Type,
         typename Pointer = std::add_pointer_t<std::add_const_t<Type>>>
-    Data<Pointer> map() const;
+    Data<Pointer> map() const &;
 
     template<
         typename Type,
         typename Pointer = std::add_pointer_t<Type>>
-    Data<Pointer> map();
+    Data<Pointer> map() &;
+
+   private:
+    // Intentionally disabed to ensure memory access is always properly
+    // encapsualted in a scoped map-unmap region.  Allowing below overloads
+    // to be invoked on a temporary would open the door to the possibility
+    // of accessing the underlying memory out of the expected scope making
+    // for seemingly ineffective memory writes and hard to hunt down bugs.
```

Comment:
I am getting that map() called on the temporary expression will be mapped-unmapped during the lifetime of that expression.

But I can not understand how that unmap will outlive that expression? Is it somewhen the chain of rvalues or something else?

Reply:
Right, great observation.  I think you have a point there and this particular application of this restriction on this class is a gray area, I'll explain why.

At a high level, and as a guideline that I try to follow, when returning the internals of an object, we want to make sure the lifetime of the returned handle (reference, pointer, etc.) will not exceed that of the object itself, so that was the high level motivation.

With that said, the ultimate usage pattern I wanted to prevent was this:

```
// Below line is a function declaration.
// Imagine we have a function that returns a `Memory` object.
Memory get_memory();

// Now we do something like this:
float * const pointer = get_memory().map<float>().get();
pointer[0] = 1.0f;

// This write above occurs outside of a mapped region.
// On top of that, if the internals of 'Memory' did not live in a pool (which it currently does)
// or if Memory would release its resources at destruction time (which it currently does not)
// then the problem was even more fundamental than access outside of a map-unmap region
// in that the pointer would now be pointing to invalid memory since 'Memory' is now destructed.
```

OK great.  Now your point made me realize that limiting `Handle::get()` to lvalue reference only would have been enough to prevent this behavior considering that the contents of `Memory` live in a pool so they outlive `Memory` anyway.  As an aside, limiting `Handle::get()` to lvalue references is something I do in this PR as well.

With that said, if we ever change `Memory` to release its contents eagerly the moment it goes out of scope, then we will need these overrides again, because `map()`'s return value provides access to the internal data.  In other words, assuming that `Memory` contents were released at the point its destructor was called, which is not the case with the current implementation by the way, then we would have:

```
// Below line is a function declaration.
// Imagine we have a function that returns a `Memory` object.
Memory get_memory();

const Memory::Data<float> data = get_memory().map();

// At this point Memory is destructed.  Data<float> holds a handle to invalid memory,
// so this usage below is problematic:

const float * pointer = data.get();
printf("%f\n", pointer[0]);
```

The reason I said this particular application of this restriction is a gray area here is because since Memory contents are managed in the pool, they outlive `Memory` so above usage is not an issue today.